// angular2-infinite-scroll/angular2-infinite-scroll.js
"use strict";
var infinite_scroll_1 = require('./src/infinite-scroll');
var position_resolver_1 = require('./src/position-resolver');
var axis_resolver_1 = require('./src/axis-resolver');
var infinite_scroll_2 = require('./src/infinite-scroll');
exports.InfiniteScroll = infinite_scroll_2.InfiniteScroll;
var position_resolver_2 = require('./src/position-resolver');
exports.PositionResolver = position_resolver_2.PositionResolver;
exports.PositionResolverFactory = position_resolver_2.PositionResolverFactory;
var axis_resolver_2 = require('./src/axis-resolver');
exports.AxisResolver = axis_resolver_2.AxisResolver;
exports.AxisResolverFactory = axis_resolver_2.AxisResolverFactory;
var index_1 = require('./src/index');
exports.InfiniteScrollModule = index_1.InfiniteScrollModule;
var scroll_register_1 = require('./src/scroll-register');
exports.ScrollRegister = scroll_register_1.ScrollRegister;
var scroll_resolver_1 = require('./src/scroll-resolver');
exports.ScrollResolver = scroll_resolver_1.ScrollResolver;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    directives: [infinite_scroll_1.InfiniteScroll, axis_resolver_1.AxisResolver, position_resolver_1.PositionResolver]
};
